transformer_engine/pytorch/csrc/extensions.h

```diff
@@ -605,3 +605,45 @@ size_t get_cudnn_version();
 bool userbuf_comm_available();
 
 void placeholder();
+
+
+/***************************************************************************************************
+ * Support THD format for Context Parallel
+ **************************************************************************************************/
+
+at::Tensor thd_read_half_tensor(const at::Tensor &tensor,
+                                const at::Tensor &cu_seqlens,
+                                int half_idx
+);
+
+void thd_lse_correction(at::Tensor &lse,
+                        const at::Tensor &lse_per_step,
+                        const at::Tensor &cu_seqlens,
+                        int total_tokens
+);
+
+at::Tensor thd_read_half_lse(const at::Tensor &lse,
```

Comment:
could this be merged with `thd_read_half_tensor`? or do we need to add `half_idx` in this function? I know we only nedd to read second half, but I think the coding is not clearly telling it's second half?

Reply:
It is not merged with `thd_read_half_tensor` for the following two reasons:
1. The shape of softmax_lse is [batch, num_heads, max_seqlen], while the shapes of q/k/v/dq/dk/dv are [total_tokens, num_heads, head_dim]. Their calculations of indices are slightly different.
2. For q, k, v, etc. For tensor in THD format, we can determine the total number of GPU threads required to launch through the size of its first dimension, but for softmax_lse, since its third dimension is max_seqlen, it requires an additional parameter ( total valid tokens) to determine how many GPU threads are launched in total